internal/controller/clientprofile_controller.go

```diff
@@ -0,0 +1,332 @@
+	// updated the config map which will result in stale representation and an update failure.
+	// The locking strategy will sync all update to the shared config file and will prevent such
+	// potential issues without a big impact on preformace as a whole
+	configMapUpdateLock.Lock()
+	defer configMapUpdateLock.Unlock()
+
+	_, err := ctrlutil.CreateOrUpdate(r.ctx, r.Client, &csiConfigMap, func() error {
+		if _, err := utils.ToggleOwnerReference(
+			!r.cleanUp,
+			&r.clientProfile,
+			&csiConfigMap,
+			r.Scheme,
+		); err != nil {
+			log.Error(err, "Failed togggling owner reference for Ceph CSI config map")
+			return err
+		}
+
+		configsAsJson := csiConfigMap.Data[utils.CsiConfigMapConfigKey]
+		clusterInfoList := []*csiClusterInfoRecord{}
+
+		// parse the json serialized list into a go array
+		if configsAsJson != "" {
+			if err := json.Unmarshal([]byte(configsAsJson), &clusterInfoList); err != nil {
+				log.Error(err, "Failed to parse cluster info list under \"config.json\" key")
+				return err
+			}
+		}
+
+		// Locate an existing entry for the same config/cluster name if exists
+		index := slices.IndexFunc(clusterInfoList, func(record *csiClusterInfoRecord) bool {
+			return record.ClusterId == r.clientProfile.Name
+		})
+
+		if !r.cleanUp {
+			// Overwrite an existing entry or append a new one
+			record := composeCsiClusterInfoRecord(&r.clientProfile, &r.cephConn)
+			if index > -1 {
+				clusterInfoList[index] = record
+			} else {
+				clusterInfoList = append(clusterInfoList, record)
+			}
+		} else if index > -1 {
+			// An O(1) unordered in-place delete of a record
+			// Will not shrink the capacity of the slice
+			length := len(clusterInfoList)
+			clusterInfoList[index] = clusterInfoList[length-1]
+			clusterInfoList = clusterInfoList[:length-1]
+		}
+
+		// Serialize the list and store it back into the config map
+		if bytes, err := json.Marshal(clusterInfoList); err != nil {
+			log.Error(err, "Failed to serialize cluster info list")
+			return err
+		} else {
+			if csiConfigMap.Data == nil {
+				csiConfigMap.Data = map[string]string{}
+			}
+			csiConfigMap.Data[utils.CsiConfigMapConfigKey] = string(bytes)
+			return nil
+		}
+	})
+
+	return err
+}
+
+// ComposeCsiClusterInfoRecord composes the desired csi cluster info record for
+// a given ClientProfile and CephConnection specs
+func composeCsiClusterInfoRecord(clientProfile *csiv1a1.ClientProfile, cephConn *csiv1a1.CephConnection) *csiClusterInfoRecord {
+	record := csiClusterInfoRecord{}
+	record.ClusterId = clientProfile.Name
+	record.Monitors = cephConn.Spec.Monitors
+	if cephFs := clientProfile.Spec.CephFs; cephFs != nil {
+		record.CephFs.SubvolumeGroup = cephFs.SubVolumeGroup
+		if mountOpt := cephFs.KernelMountOptions; mountOpt != nil {
+			record.CephFs.KernelMountOptions = utils.MapToString(mountOpt, "=", ",")
+		}
+		if mountOpt := cephFs.FuseMountOptions; mountOpt != nil {
+			record.CephFs.FuseMountOptions = utils.MapToString(mountOpt, "=", ",")
+		}
```

Comment:
i forgot the reason to make the KernelMountOptions and FuseMountOptions as map why not string?

Reply:
Some mount options are only keys, some are key-value pairs, and the operator is always fixed.

We wanted to minimize the user's chance of error in formatting the string. By providing a map interface the user does not need to concern themselves with serialization concerns.

Reply:
If we allow this we might get into a problem if the format we are passing to the ceph kernel or fuse after the operator is formatting the user input, IMO we don't need to over-simplify this for the user and the admin knows what exactly needs to be set and the operator should just set that in the configmap. my suggestion would be to make it string and pass it to the configmap.

Reply:
We agreed to shift to an array approach with a formating specifier. I will add it in a future PR.
@Madhu-1 Can we resolve this?